presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -390,8 +392,42 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
                 }
             }
 
-            // subqueries can be applied only to one side of join - left side is selected in arbitrary way
-            leftPlanBuilder = subqueryPlanner.handleUncorrelatedSubqueries(leftPlanBuilder, complexJoinExpressions, node, context);
+            if (node.getType() == LEFT || node.getType() == RIGHT) {
+                RelationType left = analysis.getOutputDescriptor(node.getLeft());
+                RelationType right = analysis.getOutputDescriptor(node.getRight());
+
+                for (Expression complexJoinExpression : complexJoinExpressions) {
+                    // IN subqueries are already disallowed previously
+                    if (subqueryPlanner.collectScalarSubqueries(complexJoinExpression, node).isEmpty() &&
+                            subqueryPlanner.collectExistsSubqueries(complexJoinExpression, node).isEmpty() &&
+                            subqueryPlanner.collectQuantifiedComparisonSubqueries(complexJoinExpression, node).isEmpty()) {
+                        continue;
+                    }
+
+                    Set<QualifiedName> dependencies = VariablesExtractor.extractNames(complexJoinExpression, analysis.getColumnReferences());
+                    if (dependencies.isEmpty() || (dependencies.stream().anyMatch(left::canResolve) && dependencies.stream().anyMatch(right::canResolve))) {
+                        // Fallback to legacy behavior
```

Comment:
a few comments/questions here:
1. i wouldn't call this falling back to legacy behavior.  It's the expected behavior for certain query shapes. 
2. Instead I would give an explanation here that because the complex join expression references both the left and right side of the join, we can't push to the right side.
3. In this case wouldn't it fail to apply the subquery just to the left side of the join since it also references the right side?

Reply:
I will reword the comment to elaborate - but essentially always attempting to push these subqueries arbitrarily into the left input does not support some query shapes that may now be supported. And it gave the "correct" behavior by accident for some query shapes. I think the expectation was that the subsequent sanity check in NoIdentifierLeftChecker would reject queries with unresolved variables, but if you look at the examples in the description of this issue it also led to obtuse error messages in some cases.

Reply:
okay, so I think your explanation of pushing to the inner side of the outer joins makes sense to me. But now I'm confused, what are the cases where we'll fail to push to the inner side of the outer join, but pushing arbitrarily to the left side of the join will succeed?  (what cases does falling back to the "legacy" behavior cover)

Reply:
I can't think of any - the only advantage is that since we call handleUncorrelatedSubqueries, but pass in the correlation, we get a [nicely formatted error message](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java#L255) as opposed to an[ illegal exception](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoIdentifierLeftChecker.java#L45)

Reply:
Any opinion on this?